src/osd/scrub_machine.cc

```diff
@@ -0,0 +1,595 @@
+
+
+using namespace std::chrono;
+using namespace std::chrono_literals;
+namespace sc = boost::statechart;
+
+#define DECLARE_LOCALS                                           \
+  ScrubMachineListener* scrbr = context<ScrubMachine>().m_scrbr; \
+  std::ignore = scrbr;                                           \
+  auto pg_id = context<ScrubMachine>().m_pg_id;                  \
+  std::ignore = pg_id;
+
+namespace Scrub {
+
+// --------- trace/debug auxiliaries -------------------------------
+
+// development code. To be removed
+void on_event_creation(std::string_view nm)
+{
+  dout(20) << " scrubberFSM event: --vvvv---- " << nm << dendl;
+}
+
+// development code. To be removed
+void on_event_discard(std::string_view nm)
+{
+  dout(20) << " scrubberFSM event: --^^^^---- " << nm << dendl;
+}
+
+void ScrubMachine::my_states() const
+{
+  for (auto si = state_begin(); si != state_end(); ++si) {
+    const auto& siw{*si};  // prevents a warning re side-effects
+    dout(20) << __func__ << " : scrub-states : " << typeid(siw).name() << dendl;
+  }
+}
+
+void ScrubMachine::assert_not_active() const
+{
+  ceph_assert(state_cast<const NotActive*>());
+}
+
+bool ScrubMachine::is_reserving() const
+{
+  return state_cast<const ReservingReplicas*>();
+}
+
+// for the rest of the code in this file - we know what PG we are dealing with:
+#undef dout_prefix
+#define dout_prefix _prefix(_dout, this->context<ScrubMachine>().m_pg)
+template <class T> static ostream& _prefix(std::ostream* _dout, T* t)
+{
+  return t->gen_prefix(*_dout) << " scrubberFSM pg(" << t->pg_id << ") ";
+}
+
+// ////////////// the actual actions
+
+// ----------------------- NotActive -----------------------------------------
+
+NotActive::NotActive(my_context ctx) : my_base(ctx)
+{
+  dout(10) << " -- state -->> NotActive" << dendl;
+}
+
+sc::result NotActive::react(const EpochChanged&)
+{
+  dout(15) << "NotActive::react(const EpochChanged&)" << dendl;
+  return discard_event();
+}
+
+// ----------------------- ReservingReplicas ---------------------------------
+
+ReservingReplicas::ReservingReplicas(my_context ctx) : my_base(ctx)
+{
+  dout(10) << " -- state -->> ReservingReplicas" << dendl;
+  DECLARE_LOCALS;  // 'scrbr' & 'pg_id' aliases
+  scrbr->reserve_replicas();
+}
+
+/**
+ *  at least one replica denied us the scrub resources we've requested
+ */
+sc::result ReservingReplicas::react(const ReservationFailure&)
+{
+  DECLARE_LOCALS;  // 'scrbr' & 'pg_id' aliases
+  dout(10) << "ReservingReplicas::react(const ReservationFailure&)" << dendl;
+
+  // the Scrubber must release all resources and abort the scrubbing
+  scrbr->clear_pgscrub_state(false);
+  return transit<NotActive>();
+}
+
+sc::result ReservingReplicas::react(const EpochChanged&)
+{
+  DECLARE_LOCALS;  // 'scrbr' & 'pg_id' aliases
+  dout(10) << "ReservingReplicas::react(const EpochChanged&)" << dendl;
+
+  // the Scrubber must release all resources and abort the scrubbing
```

Comment:
This pathway also needs to drop replica reservations.  clear_pgscrub_state currently does by resetting the ReplicaReservations object, but you'll likely need to change that.